src/posts/2020-07-30-the-worst-decision-of-my-career.md

```diff
@@ -0,0 +1,271 @@
+not the topic of today. Complexity is.
+
+## Complexity
+
+Almost everyone I know in the software industry wants to build products that
+solve real and hard problems. Change the world! If you're working on the web,
+the last thing you want is to build another CRUD application. They are fine,
+but it gets boring after a while.
+
+Most of us will get bored without novelty. That's why the software industry has
+a recycling mechanism to keep things fresh and interesting for us: every once
+in a while a new, or different, language/framework will rise and light the path
+for a brighter future, overthrowing the existing standard, demanding that we
+learn it to be on top of the game.
+
+We change our tools, but we keep building the same things over and over. I've
+been doing this long enough to see that we are just going in circles. Let's be
+honest, most of us aren't building life-changing products that wouldn't have
+been possible ten years ago, so let's not jump to the sight of another shiny
+technology that just came out and is going to solve all of our problems.
+
+## Solutions, solutions, solutions
+
+> For a while, the solution to every problem I encountered was a Rails app. --
+> _this one is mine_
+
+Over the years, I've seen companies rewrite their products to change languages,
+frameworks, or architectures because they were told that the change would make
+their product better, run faster, and scale. By the way, you're only a true
+Ruby developer after you've heard the question "but does it scale?" at least
+one hundred times (kidding, but it'll happen).
+
+Think about it, how many times were you sold a new programming language,
+technology, or a concept like microservices, serverless, event sourcing, clean
+architecture, micro frontends. There are many preachers in the software world.
+
+At Subvisual, we started using Elixir a lot these past years, so I've learned
+a bit about the history of Erlang. If you don't know, Erlang uses the Actor
+Model, which is interesting because the designers of Erlang only learned about
+the Actor Model after having designed Erlang. This is important because the
+designers of Erlang didn't start with the intent of applying the Actor Model;
+it came as a solution to fault-tolerant distributed programming.
+
+> If all you have is a hammer, everything looks like a nail _Abraham Maslow_
+
+The designers of Erlang picked the right tool for the job, and most of us want
+to do the same. Unfortunately, there's usually more than one "tool" that would
+be "right," but to know which ones, you need to know what "job" you're solving.
+All of us **should** know this, but I keep learning about teams that fail to do
+it. There are so many companies, with a handful of "experienced" developers,
+that don't know what they are building, don't have a clear business yet, but
+their product is already built on complex architectures and technologies such
+as microservices, Kubernetes or event-sourcing. I know why we make these
```

Comment:
If were bashing on things, I would add SPAs here :) Or anything that means to deal with offline as well.

Reply:
I don't think SPAs or offline are at the same level and I would rather not diverge from the main discussion.

Reply:
I think they totally are, tbh. I think that you are way more proficient with them, so you don't see it.

Reply:
Not what I was trying to do. What I meant was that when you are comfortable with a technology, it seems simpler to you than it probably is to beginners. It happens a lot with the "Rails magic" thing, I don't see it as magic because I'm so familiar with how things work that it makes total sense to me.

Reply:
I don't want to add SPAs and offline to this list because it makes no sense to me. I'm usually more vocal regarding front-end stuff, but it isn't even what I have more experience with: I spent the last two years working in Elixir and dealing with ES issues, DevOps and others; and before that, I was working with Ruby on Nourish's events system, where I used materialized views and triggers and all of that. Sure, I'm not that experienced in microservices and Kubernetes, but I'm not a beginner because of the experience I have on the technologies around those. We'll just go in circles, I know you have a special hate for JS and SPAs and all of that, but I really don't want to have anything to do with it in this blog post.

Reply:
Well, I think the opposite. They are all examples of overly complex things people tend to reach for wait too early in projects, in my opinion. But let's agree to disagree, then.